src/formats/pnm.c

```diff
@@ -0,0 +1,222 @@
+// SPDX-License-Identifier: MIT
+// PNM formats decoder
+// Copyright (C) 2023 Abe Wieland <abe.wieland@gmail.com>
+
+#include "loader.h"
+
+#include <setjmp.h>
+
+// Divide a by b, rounding to the nearest integer; evaluates b twice
+#define div_near(a, b) (((a) + (b) / 2) / (b));
+// Divide a by b, rounding up; evalutes b twice
+#define div_ceil(a, b) (((a) + (b) - 1) / (b));
+
+// A file-like abstraction with error handling via setjmp/longjmp which
+// significantly helps code readability
+struct pnm_iter {
+    const uint8_t* data;
+    size_t pos;
+    size_t size;
+    jmp_buf jmp;
+};
+
+// Error conditions as passed to longjmp
+#define ERR_EOF     1
+#define ERR_RANGE   2
+#define ERR_NONDIG  3
+
+// Read a character, ignoring comments
+static uint8_t pnm_readc(struct pnm_iter* it)
+{
+    if (it->pos == it->size) longjmp(it->jmp, ERR_EOF);
+    if (it->data[it->pos] == '#') {
+        do {
+            ++it->pos;
+        } while (it->pos < it->size && it->data[it->pos] != '\n'
+                 && it->data[it->pos] != '\r');
+        if (it->pos == it->size) longjmp(it->jmp, ERR_EOF);
+    }
+    return it->data[it->pos++];
+}
+
+// Read an integer, ignoring leading whitespace
+static size_t pnm_readuint(struct pnm_iter* it)
+{
+    uint8_t c;
+    do {
+        c = pnm_readc(it);
+    } while (c == ' ' || c == '\t' || c == '\n' || c == '\r');
+    if (c < '0' || c > '9') longjmp(it->jmp, ERR_NONDIG);
+    size_t val = 0;
+    do {
+        uint8_t d = c - '0';
+        if (val > SIZE_MAX / 10) longjmp(it->jmp, ERR_RANGE);
+        val *= 10;
+        if (val > SIZE_MAX - d) longjmp(it->jmp, ERR_RANGE);
+        val += d;
+        c = pnm_readc(it);
+    } while (c >= '0' && c <= '9');
+    --it->pos;
+    return val;
+}
+
+// PNM loader implementation
+enum loader_status decode_pnm(struct image* ctx, const uint8_t* data,
+                              size_t size)
+{
+    bool plain;
+    int type;
+    struct pnm_iter it;
+    size_t width, height, maxval;
+
+    if (size < 2 || data[0] != 'P' || data[1] < '1' || data[1] > '6') {
+        return ldr_unsupported;
+    }
+    plain = data[1] < '4';
+    type = (data[1] - '1') % 3;
```

Comment:
Does the PNM specification really uses `%` to specify the type? =)

Reply:
Well, `P1` and `P4` are both PBM files, `P2` and `P5` are both PGM, and `P3` and `P6` are both PPM files. The lower number for each is the ASCII/plain representation, while the higher number is the raw/binary representation. If you prefer, I could definitely write something like
```
#define PBM_PLAIN '1'
...
#define PPM_RAW   '6'
...
#define TYPE_PBM   0
#define TYPE_PGM   1
#define TYPE_PPM   2
...
switch (data[1]) {
    case PBM_PLAIN:
        plain = true;
        type = TYPE_PBM;
        break;
    ....
    case PPM_RAW:
        plain = false;
        type = TYPE_PPM;
        break;
}
```